### PR TITLE
feat: Experimental `credential_provider` argument for `scan_parquet`

### PR DESCRIPTION
**Pull request by nameexhaustion**
ref [Allow Overriding Object Store Credential Provider #18979](https://github.com/pola-rs/polars/issues/18979)

Adds an experimental `credential_provider` argument to `scan_parquet`, which can accept a callable:

```python
def credential_provider() -> tuple[dict[str, str | None], int | None]
```

Where the dictionary contains the credentials, and the optional 2nd value represents the expiry time as seconds since unix epoch, allowing us to avoid calling the function again until the credentials have expired (if it is `None`, then the credentials will be interpreted as to never expire).

This enables using custom credential provisioning logic with cloud I/O scans, which allows for handling credential expiry.

#### Example for AWS
```python

def get_credentials():
    import boto3

    session = boto3.Session(profile_name="profile2")
    creds = session.get_credentials()

    return {
        "aws_access_key_id": creds.access_key,
        "aws_secret_access_key": creds.secret_key,
        "aws_session_token": creds.token,
    }, None


lf = pl.scan_parquet(
    "s3://...",
    credential_provider=get_credentials,
)

```

#### Example for GCP
```python

def get_credentials():
    from pathlib import Path

    import google.auth
    import google.auth.transport.requests
    import zoneinfo

    creds, _ = google.auth.load_credentials_from_file(
        Path.home() / ".config/gcloud/application_default_credentials.json"
    )

    auth_req = google.auth.transport.requests.Request()
    creds.refresh(auth_req)

    return {"bearer_token": creds.token}, int(
        creds.expiry.replace(
            # Google auth does not set this properly
            tzinfo=zoneinfo.ZoneInfo("UTC")
        ).timestamp()
    )


lf = pl.scan_parquet(
    "gs://...",
    credential_provider=get_credentials,
)

```

#### Azure
I don't have an Azure environment to test on, but the returned dictionary should contain `{ 'bearer_token': '...' }`

Todos for follow-up PRs
* [ ] Add the option to other cloud-enabled `scan_/read_` functions
* [ ] Add and link to a documentation page on the function return formats for different cloud types

### CODE REVIEW

File: crates/polars-plan/src/dsl/python_udf.rs [hunk 1]
```diff
@@ -26,88 +21,10 @@ pub static mut CALL_COLUMNS_UDF_PYTHON: Option<
 pub static mut CALL_DF_UDF_PYTHON: Option<
     fn(s: DataFrame, lambda: &PyObject) -> PolarsResult<DataFrame>,
 > = None;
-#[cfg(feature = "serde")]
-pub(super) const MAGIC_BYTE_MARK: &[u8] = "PLPYUDF".as_bytes();
-static PYTHON_VERSION_MINOR: Lazy<u8> = Lazy::new(get_python_minor_version);
 
-#[derive(Debug)]
-pub struct PythonFunction(pub PyObject);
```
Comment (nameexhaustion) [inline crates/polars-plan/src/dsl/python_udf.rs:34]: I have moved `PythonFunction` from `polars-plan/../python_udf.rs` to `polars-utils` for re-use

File: crates/polars-io/src/cloud/credential_provider.rs [hunk 2]
```diff
@@ -0,0 +1,679 @@
...
+            return v.serialize(serializer);
+        }
+
+        Err(S::Error::custom(format!("cannot serialize {:?}", self)))
+    }
+}
+
+/// Avoids calling the credential provider function if we have not yet passed the expiry time.
+#[derive(Debug)]
+struct FetchedCredentialsCache<C>(tokio::sync::Mutex<(C, u64)>);
```
Comment (ritchie46) [inline crates/polars-io/src/cloud/credential_provider.rs:352]: Nice one!

File: py-polars/polars/io/parquet/functions.py [hunk 3]
```diff
@@ -467,6 +476,10 @@ def scan_parquet(
         msg = "The `hive_schema` parameter of `scan_parquet` is considered unstable."
         issue_unstable_warning(msg)
 
+    if credential_provider is not None:
```
Comment (ritchie46) [inline py-polars/polars/io/parquet/functions.py:479]: Can we if the `credential_provider` and `storage_options` are `None`, the link is a cloud url and boto3/gcp is installed, insert a `credential_provider` function on behalf of the user?

File: crates/polars-io/src/cloud/credential_provider.rs (hunk 2 above)
Comment (nameexhaustion) [reply crates/polars-io/src/cloud/credential_provider.rs:352]: I tried changing it but couldn't as `object_store::CredentialProvider` requires `Send`, so we have to use tokio's mutex

File: py-polars/polars/io/parquet/functions.py (hunk 3 above)
Comment (nameexhaustion) [reply py-polars/polars/io/parquet/functions.py:479]: We could, but then I think there wouldn't be any way for users to disable that behavior, so I'm not sure.

I'm thinking in a follow-up PR we can add something that lets users do

```python
pl.scan_parquet(..., credential_provider=pl.CredentialProviderAWS(sts_role=..., profile_name=...))
```

Comment (ritchie46) [reply py-polars/polars/io/parquet/functions.py:479]: But ok, let's do that in a follow up PR.

**Pull request merged**